File: mvp/src/main/java/me/giacoppo/mvp/repository/RepositoryPattern.java
package me.giacoppo.mvp.repository;

import java.util.List;

import io.reactivex.Observable;

public interface RepositoryPattern<T> {
    Observable<Void> add(T item);
    Observable<Void> add(Iterable<T> items);
    Observable<Void> update(T item);
    Observable<Void> remove(T item);
    Observable<Void> remove(Specification specification);
    Observable<List<T>> query(Specification specification);
}

File: mvp/src/main/java/me/giacoppo/mvp/repository/RepositoryPattern.kt
package me.giacoppo.mvp.repository

import io.reactivex.Observable

interface RepositoryPattern<T, S : Specification> {
    fun <X> add(item: T): Observable<X>
    fun <X> add(items: Iterable<T>): Observable<X>
    fun <X> update(item: T): Observable<X>
    fun <X> remove(item: T): Observable<X>
    fun <X> remove(specification: S): Observable<X>
    operator fun get(specification: S): Observable<List<T>>
}
